test/GraphQL.Conventions.Tests/Adapters/Engine/ErrorTests.cs

```diff
@@ -76,51 +77,56 @@ public async Task Will_Provide_Path_And_Code_For_Errors_In_Array_Fields()
             var result = await engine
                 .NewExecutor()
                 .WithQueryString("query Blah { getObject { arrayField { test } } }")
-                .Execute();
-
+                .ExecuteAsync();
+            
             result.Data.ShouldHaveFieldWithValue("getObject", "arrayField", 0, "test", "some value");
             result.Data.ShouldHaveFieldWithValue("getObject", "arrayField", 1, "test", null);
             result.Data.ShouldHaveFieldWithValue("getObject", "arrayField", 2, "test", "some value");
             result.Data.ShouldHaveFieldWithValue("getObject", "arrayField", 3, "test", null);
-
+            
             result.Errors.ShouldNotBeNull();
             result.Errors.Count.ShouldEqual(2);
 
             var error = result.Errors.ElementAt(0);
             error.Message.ShouldEqual("Test error.");
-            error.Code.ShouldEqual("ARGUMENT");
+            error.Code.ShouldEqual("FIELD_RESOLUTION");
             error.Path.Count().ShouldEqual(4);
             error.Path.ElementAt(0).ShouldEqual("getObject");
             error.Path.ElementAt(1).ShouldEqual("arrayField");
-            error.Path.ElementAt(2).ShouldEqual("1");
+            error.Path.ElementAt(2).ShouldEqual(1);
             error.Path.ElementAt(3).ShouldEqual("test");
 
             error = result.Errors.ElementAt(1);
             error.Message.ShouldEqual("Test error.");
-            error.Code.ShouldEqual("ARGUMENT");
+            error.Code.ShouldEqual("FIELD_RESOLUTION");
             error.Path.Count().ShouldEqual(4);
             error.Path.ElementAt(0).ShouldEqual("getObject");
             error.Path.ElementAt(1).ShouldEqual("arrayField");
-            error.Path.ElementAt(2).ShouldEqual("3");
+            error.Path.ElementAt(2).ShouldEqual(3);
             error.Path.ElementAt(3).ShouldEqual("test");
         }
 
         [Test]
         public async Task Will_Provide_Exception_Data()
         {
             var engine = GraphQLEngine.New<Query>();
+
             var result = await engine
                 .NewExecutor()
                 .WithQueryString("query Blah { errorWithData }")
-                .Execute();
+                .ExecuteAsync();
 
             result.Data.ShouldHaveFieldWithValue("errorWithData", null);
-
             result.Errors.ShouldNotBeNull();
             result.Errors.Count.ShouldEqual(1);
 
-            result.Errors.First().Message.ShouldEqual("Test error.");
-            result.Errors.First().Data["someKey"].ShouldEqual("someValue");
+            var error = result.Errors.First();
+            error.InnerException.ShouldNotBeNull();
+            error.InnerException.InnerException.ShouldNotBeNull();
```

Comment:
What's the cause for the nesting here...? What's changed, and how does this error actually manifest itself in the response payload now?

Reply:
The source exception is wrapped twice in graphql-dotnet. Once by the .net framework (methodInfo.Invoke) and once by the ExecutionStrategy (ProcessNodeUnhandledException).

Reply:
Gotcha. Sounds like we can remove at least one level of indirection there, no?

Reply:
Not to access the data attached to the exception. We could unwrap the exception in GraphQLEngine (via an ErrorTransformer maybe) but it would probably be best to handle that in Graphql.net directly as it is the one who wraps it.

Reply:
Ah, I see - let me take it for a spin later and see for myself what the current behaviour is. Will get back to you.

Reply:
I was writing code earlier today to prevent `MethodInfo.Invoke` from wrapping an exception.  I explored and tested `MethodInfo.Invoke` and `Delegate.DynamicInvoke`, and explored using `ExceptionDispatchInfo.Capture`.  I settled on compiling an expression so there is no need to rethrow the exception, using a cache to reduce the impact of dynamically compiling the expression.  I can paste the code in here for you to use @schabannes @tlil .  Using `ExceptionDispatchInfo.Capture` is probably the best alternative choice if this seems too complex.

Reply:
Yes, please @Shane32 - if you could share, I'd be happy to take a look.

Reply:
```csharp
    class MethodInvokeTests
    {
        public async Task RunAsync()
        {
            Console.WriteLine($"StaticTest1: {Invoke(GetMethodInfo("StaticTest1"), null, null)}");
            Console.WriteLine($"StaticTest2: {Invoke(GetMethodInfo("StaticTest2"), null, null)}");
            Console.WriteLine($"StaticTest3: {Invoke(GetMethodInfo("StaticTest3"), null, null)}");
            Console.WriteLine($"InstanceTest1: {Invoke(GetMethodInfo("InstanceTest1"), this, null)}");
            Console.WriteLine($"InstanceTest2: {Invoke(GetMethodInfo("InstanceTest2"), this, null)}");
            Console.WriteLine($"InstanceTest3: {Invoke(GetMethodInfo("InstanceTest3"), this, null)}");
            var guid = Guid.NewGuid();
            var stream = new System.IO.MemoryStream();
            var value = (int?)35;
            Console.WriteLine($"StaticWithParams: {Invoke(GetMethodInfo("StaticWithParams"), null, new object[] { guid, stream, value })}");
            Console.WriteLine($"InstanceWithParams: {Invoke(GetMethodInfo("InstanceWithParams"), this, new object[] { guid, stream, value })}");
        }

        private MethodInfo GetMethodInfo(string name) => typeof(MethodInvokeTests).GetMethod(name, BindingFlags.Public | BindingFlags.NonPublic | BindingFlags.Instance | BindingFlags.Static);
        private static int StaticTest1() => 2;
        private static string StaticTest2() => "hello";
        private static void StaticTest3() => Console.WriteLine("ran StaticTest3");
        private int InstanceTest1() => 3;
        private string InstanceTest2() => "hello2";
        private void InstanceTest3() => Console.WriteLine("ran InstanceTest3");
        private static string StaticWithParams(Guid guid, Stream stream, int? value) => $"Static Received {guid} {stream} {value}";
        private string InstanceWithParams(Guid guid, Stream stream, int? value) => $"Instance Received {guid} {stream} {value}";

        private static ConcurrentDictionary<MethodInfo, Func<object, object[], object>> _methodDictionary = new ConcurrentDictionary<MethodInfo, Func<object, object[], object>>();
        public static object Invoke(MethodInfo methodInfo, object instance, params object[] arguments)
        {
            //just good practice
            if (methodInfo == null)
                throw new ArgumentNullException(nameof(methodInfo));
            //the following check is ignored by default
            if (methodInfo.IsStatic && instance != null)
                throw new ArgumentException("Instance must be null for static methods", nameof(instance));
            //the lambda ignores extra arguments so the following check is necessary
            if (methodInfo.GetParameters().Length != (arguments?.Length ?? 0))
                throw new ArgumentException("Invalid number of arguments for this method", "arguments");
            //type errors will be thrown as InvalidCastException inside the lambda
            //null instance (for non-static methods) will be thrown as NullReferenceException inside the lambda

            var lambda = _methodDictionary.GetOrAdd(methodInfo, (methodInfo) => {
                var instanceParameter = Expression.Parameter(typeof(object));
                var argumentsParameter = Expression.Parameter(typeof(object[]));
                var instanceExpression = methodInfo.IsStatic ? null : Expression.Convert(instanceParameter, methodInfo.DeclaringType);
                var parameters = methodInfo.GetParameters().Select((param, index) => {
                    return Expression.Convert(Expression.ArrayAccess(argumentsParameter, Expression.Constant(index)), param.ParameterType);
                });
                var call = Expression.Call(instanceExpression, methodInfo, parameters);
                Expression body;
                if (methodInfo.ReturnType == typeof(void)) {
                    body = Expression.Block(new Expression[] {
                        call,
                        Expression.Constant(null, typeof(object))
                    });
                }
                else {
                    body = Expression.Convert(call, typeof(object));
                }
                var lambda = Expression.Lambda<Func<object, object[], object>>(body, instanceParameter, argumentsParameter);
                return lambda.Compile();
            });

            return lambda(instance, arguments);
        }
    }
```

Reply:
And here's the simple version, which does almost the same thing, catching and rethrowing the inner exception while maintaining the stack trace:
```csharp
        public static object Invoke2(MethodInfo methodInfo, object instance, params object[] arguments)
        {
            try {
                return methodInfo.Invoke(instance, arguments);
            }
            catch (TargetInvocationException ex) {
                ExceptionDispatchInfo.Capture(ex.InnerException).Throw();
                return null; //does not run, but intellisense doesn't know that
            }
        }
```
Note that in theory, this could catch a `TargetInvocationException` that occurred within the method.

In terms of performance, the former option is probably slower for the first-time execution, as it has to compile an expression into a dynamically generated assembly at runtime.  But it should be just as fast after the first time; perhaps even faster.  It would be interesting to see performance benchmarks and see which is faster in the long run.

Personally, I don't like catching and rethrowing exceptions, as when debugging an unhandled exception, Visual Studio invariably seems to highlight the line of code that's rethrowing the exception, not where it originally occurred.  Then, and the stack traces don't contain a "--- End of stack trace from previous location where exception was thrown ---" message.  Besides those reasons, the simple alternative is probably perfectly adequate.

Reply:
I've not been following the discussion so I'm not sure where this is used, and so perhaps my code could be tweaked for better performance given the usage.  Also I could have commented it better and added xml comments...  (Actually, xml comments could be copied and pasted from MethodInfo.Invoke, I bet.)

Reply:
Nice. I'll try to wrap that in a PR on the graphql-dotnet main repo.

Reply:
@schabannes Actually, the main repo already unwraps exceptions triggered by MethodInfo.Invoke.  It does not, however, retain the proper stack trace or unwrap exceptions triggered by ConstructorInfo.Invoke.  I've proposed slight changes in https://github.com/graphql-dotnet/graphql-dotnet/pull/1988 .  I've also put together PR #210 which contains the necessary changes to this project.  As this project relies heavily on `MethodInfo.Invoke`, I've included the more exotic code so that exceptions are not rethrown at all, along with extensive tests.